bin/grade_item.py

```diff
@@ -0,0 +1,412 @@
+                     "{}",
+                     ";"])
+
+    # --------------------------------------------------------------------
+    # RUN VALIDATOR
+
+    with open(os.path.join(tmp_logs,"overall.txt"),'a') as f:
+        print ("====================================\nVALIDATION STARTS", file=f)
+
+    # copy results files from compilation...
+    patterns_submission_to_validation = complete_config_obj["autograding"]["submission_to_validation"]
+    pattern_copy("submission_to_validation",patterns_submission_to_validation,submission_path,tmp_work,tmp_logs)
+    patterns_compilation_to_validation = complete_config_obj["autograding"]["compilation_to_validation"]
+    pattern_copy("compilation_to_validation",patterns_compilation_to_validation,tmp_compilation,tmp_work,tmp_logs)
+    
+    # remove the compilation directory
+    shutil.rmtree(tmp_compilation)
+
+    # copy output files to tmp_work directory
+    copy_contents_into(test_output_path,tmp_work)
+
+    subprocess.call("ls -la "+tmp_work+" >> "+tmp_logs+"/overall.txt",shell=True)
+        
+    # copy validator.out to the current directory
+    shutil.copy (os.path.join(bin_path,obj["gradeable"],"validate.out"),os.path.join(tmp_work,"my_validator.out"))
+
+    # give the untrusted user read/write/execute permissions on the tmp directory & files
+    add_permissions_recursive(tmp_work,
+                              stat.S_IROTH | stat.S_IWOTH | stat.S_IXOTH,
+                              stat.S_IROTH | stat.S_IWOTH | stat.S_IXOTH,
+                              stat.S_IROTH)
+
+    add_permissions(os.path.join(tmp_work,"my_validator.out"),stat.S_IROTH | stat.S_IXOTH)
+
+    # validator the validator.out as the untrusted user
+    with open(os.path.join(tmp_logs,"validator_log.txt"), 'w') as logfile:
+        validator_success = subprocess.call([os.path.join(SUBMITTY_INSTALL_DIR,"bin","untrusted_execute"),
+                                             args.which_untrusted,
+                                             os.path.join(tmp_work,"my_validator.out"),
+                                             obj["gradeable"],
+                                             obj["who"],
+                                             str(obj["version"]),
+                                             submission_time],
+                                            stdout=logfile)
+
+    if validator_success == 0:
+        print ("NEW VALIDATOR OK")
+    else:
+        print ("NEW VALIDATOR FAILURE")
+
+    # give permissions to all created files to the hwcron user
+    subprocess.call([os.path.join(SUBMITTY_INSTALL_DIR,"bin","untrusted_execute"),
+                     args.which_untrusted,
+                     "/usr/bin/find",
+                     tmp_work,
+                     "-user",
+                     args.which_untrusted,
+                     "-exec",
+                     "/bin/chmod",
+                     "o+r",
+                     "{}",
+                     ";"])
```

Comment:
This is the third time I've seen this command structure. Can we make this a function that just takes in a directory since that appears to be only the difference between them and it's a somewhat complicated bit of logic.